test/spmd/test_xla_distributed_checkpoint.py

```diff
@@ -403,6 +403,44 @@ def test_manager_max_to_keep(self, tmpdir):
     self.assertTrue(chkpt_mgr.save(30, state_dict))
     self.assertEqual(set(chkpt_mgr.all_steps()), {30, 20})
 
+  @run_with_tmpdir
+  def test_manager_async_checkpoint(self, tmpdir):
+    chkpt_mgr = CheckpointManager(tmpdir, save_period=10)
+    state_dict = self._get_sharded_model().state_dict()
+
+    self.assertEqual(chkpt_mgr.all_steps(), [])
+
+    # Steps not divisible by 10 should not be saved
+    for step in range(1, 10):
+      self.assertFalse(chkpt_mgr.save_async(step, state_dict))
+      self.assertEqual(chkpt_mgr.all_steps(), [])
+
+    # Steps divisible by 10 should be saved
+    saved = set()
+    for step in range(0, 100, 10):
+      self.assertTrue(chkpt_mgr.save_async(step, state_dict))
+      saved.add(step)
+
+    # Delete the checkpoint manager to block this thread until all pending
+    # async checkpoints are complete.
+    del chkpt_mgr
```

Comment:
Not sure if this can test the saving is async but not sync?

Reply:
I'm worried about introducing flakiness to the tests... Do you have any thoughts on how to assert it's an async operation? Maybe I can add some counters.

Reply:
We can mock it in a way that the exec function in the manager never return until a signal sent later in your test after a check.

Reply:
Yea, this is concise and nice, though. it deduces it from the number of saved chkpts and threading.Thread(target=self._async_worker, daemon=True) did the job with the ending `_async_queue.join()` call.

@jonb377 it looks like you are spawning a single-thread per manager, how do you control the number of threads?

Reply:
That's a great callout, it should be addressed in the next patch. I'm going to allow the threads to exit when the manager is deleted.

Reply:
also, if you can't rely on _del_ being called, you may wanto throw in `await`.